message/v1/message.go

```diff
@@ -0,0 +1,315 @@
+		})
+	}
+
+	blocks := gsm.Blocks()
+	pbm.Data = make([]*pb.Message_V1_0_0_Block, 0, len(blocks))
+	for _, b := range blocks {
+		pbm.Data = append(pbm.Data, &pb.Message_V1_0_0_Block{
+			Prefix: b.Cid().Prefix().Bytes(),
+			Data:   b.RawData(),
+		})
+	}
+	return pbm, nil
+}
+
+// Mapping from a pb.Message_V1_0_0 object to a GraphSyncMessage object, including
+// RequestID (int / uuid) mapping.
+func (mh *MessageHandler) newMessageFromProto(p peer.ID, pbm *pb.Message_V1_0_0) (message.GraphSyncMessage, error) {
+	mh.mapLock.Lock()
+	defer mh.mapLock.Unlock()
+
+	requests := make(map[graphsync.RequestID]message.GraphSyncRequest, len(pbm.GetRequests()))
+	for _, req := range pbm.Requests {
+		if req == nil {
+			return message.GraphSyncMessage{}, errors.New("request is nil")
+		}
+		var err error
+
+		id, err := intIdToRequestId(p, mh.fromV1Map, mh.toV1Map, req.Id)
+		if err != nil {
+			return message.GraphSyncMessage{}, err
+		}
+
+		if req.Cancel {
+			requests[id] = message.NewCancelRequest(id)
+			continue
+		}
+
+		exts, err := fromEncodedExtensions(req.GetExtensions())
+		if err != nil {
+			return message.GraphSyncMessage{}, err
+		}
+
+		if req.Update {
+			requests[id] = message.NewUpdateRequest(id, exts...)
+			continue
+		}
+
+		root, err := cid.Cast(req.Root)
+		if err != nil {
+			return message.GraphSyncMessage{}, err
+		}
+
+		selector, err := ipldutil.DecodeNode(req.Selector)
+		if err != nil {
+			return message.GraphSyncMessage{}, err
+		}
+
+		requests[id] = message.NewRequest(id, root, selector, graphsync.Priority(req.Priority), exts...)
+	}
+
+	responses := make(map[graphsync.RequestID]message.GraphSyncResponse, len(pbm.GetResponses()))
+	for _, res := range pbm.Responses {
+		if res == nil {
+			return message.GraphSyncMessage{}, errors.New("response is nil")
+		}
+		id, err := intIdToRequestId(p, mh.fromV1Map, mh.toV1Map, res.Id)
+		if err != nil {
+			return message.GraphSyncMessage{}, err
+		}
+		exts, err := fromEncodedExtensions(res.GetExtensions())
+		if err != nil {
+			return message.GraphSyncMessage{}, err
+		}
+		responses[id] = message.NewResponse(id, graphsync.ResponseStatusCode(res.Status), exts...)
+	}
+
+	blks := make(map[cid.Cid]blocks.Block, len(pbm.GetData()))
+	for _, b := range pbm.GetData() {
+		if b == nil {
+			return message.GraphSyncMessage{}, errors.New("block is nil")
+		}
+
+		pref, err := cid.PrefixFromBytes(b.GetPrefix())
+		if err != nil {
+			return message.GraphSyncMessage{}, err
+		}
+
+		c, err := pref.Sum(b.GetData())
+		if err != nil {
+			return message.GraphSyncMessage{}, err
+		}
+
+		blk, err := blocks.NewBlockWithCid(b.GetData(), c)
+		if err != nil {
+			return message.GraphSyncMessage{}, err
+		}
+
+		blks[blk.Cid()] = blk
+	}
+
+	return message.NewMessage(requests, responses, blks), nil
+}
+
+// TODO: is this a breaking protocol change? force all extension data into dag-cbor?
+func toEncodedExtensions(part MessagePartWithExtensions) (map[string][]byte, error) {
```

Comment:
The assumptions around `toEncodedExtensions ` and `fromEncodedExtensions` are my major concerns wrt breakage at the moment (see TODOs).

It forces all extensions to be dag-cbor bytes, which might be fine for users of this library since they are forced to hand us `datamodel.Node`s for the extension data from now on, but when we're talking to v1 clients, we'll receive their extensions as bytes in the protbuf and decode them as dag-cbor, erroring if they don't decode. That means that if the remote peer is speaking v1 then we can only receive dag-cbor as extension data and will error in any other case.

Is there are case to do a try-decode and fall back to wrapping the raw bytes in a `basicnode.NewBytes()` so we retain the data and make it up to the library user to deal with those raw bytes (perhaps they have their own protobuf format to decode them into?).